README.md
# itunes-feed
[![npm](https://img.shields.io/npm/v/itunes-feed.svg?style=for-the-badge)](https://www.npmjs.com/package/itunes-feed)
[![npm](https://img.shields.io/npm/dt/itunes-feed.svg?style=for-the-badge)](https://www.npmjs.com/package/itunes-feed)
[![David](https://img.shields.io/david/KyleRoss/node-itunes-feed.svg?style=for-the-badge)](https://david-dm.org/KyleRoss/node-itunes-feed)
[![Travis](https://img.shields.io/travis/KyleRoss/node-itunes-feed/master.svg?style=for-the-badge)](https://travis-ci.org/KyleRoss/node-itunes-feed)
[![Coveralls](https://img.shields.io/coveralls/github/KyleRoss/node-itunes-feed.svg?style=for-the-badge)](https://coveralls.io/github/KyleRoss/node-itunes-feed)
[![license](https://img.shields.io/github/license/KyleRoss/node-itunes-feed.svg?style=for-the-badge)](https://github.com/KyleRoss/node-itunes-feed/blob/master/LICENSE)
[![Beerpay](https://img.shields.io/beerpay/KyleRoss/node-itunes-feed.svg?style=for-the-badge)](https://beerpay.io/KyleRoss/node-itunes-feed)

Generates and retrieves JSON feeds from iTunes for various different services. A programmatic version of the [RSS Feed Generator](https://rss.itunes.apple.com/en-us).

# Usage
## Install
```
npm install --save itunes-feed
```

## Example
```js
const itunesFeed = require('itunes-feed');

// ...

try {
    let feed = await itunesFeed.appleMusic();
    console.log(`Retrieved ${feed.length} items for Apple Music`);
    
    // ...
} catch(error) {
    console.error(error);
}
```

# API Documentation
<a name="module_itunes-feed"></a>

## itunes-feed ⇒ [<code>ITunesFeed</code>](#ITunesFeed)
**Returns**: [<code>ITunesFeed</code>](#ITunesFeed) - Instance of ITunesFeed.  
**Example**  
```js
// Basic usage
const itunesFeed = require('itunes-feed');

// Advanced usage
const ITunesFeed = require('itunes-feed').ITunesFeed;
const itunesFeed = new ITunesFeed({
    // options...
});
```
<a name="ITunesFeed"></a>

## ITunesFeed
Class exported from itunes-feed.

**Kind**: global class  

* [ITunesFeed](#ITunesFeed)
    * [new ITunesFeed([options])](#new_ITunesFeed_new)
    * [itunesFeed.ITunesFeed](#ITunesFeed+ITunesFeed) : [<code>ITunesFeed</code>](#ITunesFeed)
    * [itunesFeed.options](#ITunesFeed+options) : <code>Object</code>
    * [itunesFeed.appleMusic([feedType], [options])](#ITunesFeed+appleMusic) ⇒ [<code>Promise.&lt;Feed&gt;</code>](#Feed)
    * [itunesFeed.audioBooks([options])](#ITunesFeed+audioBooks) ⇒ [<code>Promise.&lt;Feed&gt;</code>](#Feed)
    * [itunesFeed.books([feedType], [options])](#ITunesFeed+books) ⇒ [<code>Promise.&lt;Feed&gt;</code>](#Feed)
    * [itunesFeed.iosApps([feedType], [genre], [options])](#ITunesFeed+iosApps) ⇒ [<code>Promise.&lt;Feed&gt;</code>](#Feed)
    * [itunesFeed.itunesMusic([feedType], [options])](#ITunesFeed+itunesMusic) ⇒ [<code>Promise.&lt;Feed&gt;</code>](#Feed)
    * [itunesFeed.itunesU([options])](#ITunesFeed+itunesU) ⇒ [<code>Promise.&lt;Feed&gt;</code>](#Feed)
    * [itunesFeed.macApps([feedType], [options])](#ITunesFeed+macApps) ⇒ [<code>Promise.&lt;Feed&gt;</code>](#Feed)
    * [itunesFeed.movies([options])](#ITunesFeed+movies) ⇒ [<code>Promise.&lt;Feed&gt;</code>](#Feed)
    * [itunesFeed.musicVideos([options])](#ITunesFeed+musicVideos) ⇒ [<code>Promise.&lt;Feed&gt;</code>](#Feed)
    * [itunesFeed.podcasts([options])](#ITunesFeed+podcasts) ⇒ [<code>Promise.&lt;Feed&gt;</code>](#Feed)
    * [itunesFeed.tvShows([feedType], [options])](#ITunesFeed+tvShows) ⇒ [<code>Promise.&lt;Feed&gt;</code>](#Feed)

<a name="new_ITunesFeed_new"></a>

### new ITunesFeed([options])
Creates a new instance of ITunesFeed.


| Param | Type | Default | Description |
| --- | --- | --- | --- |
| [options] | <code>Object</code> | <code>{}</code> | Optional object for configuring ITunesFeed. These options can be set on the instance (`this.options`). |
| [options.countryCode] | <code>String</code> | <code>&#x27;us&#x27;</code> | 2 letter country code to use for all requests. Must be lowercase. |
| [options.limit] | <code>Number</code> | <code>100</code> | Limit of items returned in a feed. Must be one of: `10`, `25`, `50` or `100`. Any other value will throw an error. |
| [options.explicit] | <code>Boolean</code> | <code>true</code> | Whether to include explicit results. |
| [options.axiosOptions] | <code>Object</code> |  | Object containing additional options to pass to [axios](https://github.com/axios/axios#request-config). |

<a name="ITunesFeed+ITunesFeed"></a>

### itunesFeed.ITunesFeed : [<code>ITunesFeed</code>](#ITunesFeed)
Access to uninstantiated ITunesFeed class for advanced usage.

**Kind**: instance property of [<code>ITunesFeed</code>](#ITunesFeed)  
**Read only**: true  
<a name="ITunesFeed+options"></a>

### itunesFeed.options : <code>Object</code>
Global options set on this instance. You may change these options at any time.

**Kind**: instance property of [<code>ITunesFeed</code>](#ITunesFeed)  
**Example**  
```js
itunesFeed.options.limit = 50;
```
<a name="ITunesFeed+appleMusic"></a>

### itunesFeed.appleMusic([feedType], [options]) ⇒ [<code>Promise.&lt;Feed&gt;</code>](#Feed)
Retrieves feed for Apple Music.

**Kind**: instance method of [<code>ITunesFeed</code>](#ITunesFeed)  
**Returns**: [<code>Promise.&lt;Feed&gt;</code>](#Feed) - Instance of `Feed`.  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| [feedType] | <code>String</code> | <code>&#x27;top-songs&#x27;</code> | The type of feed to retrieve. Must be one of: `coming-soon`, `hot-tracks`, `new-releases`, `top-albums`, `top-songs`. |
| [options] | <code>Object</code> | <code>{}</code> | Any options from ITunesFeed to overwrite for this request. |

**Example**  
```js
let feed = await itunesFeed.appleMusic();
//=> [...]

let feed = await itunesFeed.appleMusic('hot-tracks', {
    explicit: false
});
```
<a name="ITunesFeed+audioBooks"></a>

### itunesFeed.audioBooks([options]) ⇒ [<code>Promise.&lt;Feed&gt;</code>](#Feed)
Retrieves feed for Audiobooks.

**Kind**: instance method of [<code>ITunesFeed</code>](#ITunesFeed)  
**Returns**: [<code>Promise.&lt;Feed&gt;</code>](#Feed) - Instance of `Feed`.  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| [options] | <code>Object</code> | <code>{}</code> | Any options from ITunesFeed to overwrite for this request. |

**Example**  
```js
let feed = await itunesFeed.audioBooks();
//=> [...]
```
<a name="ITunesFeed+books"></a>

### itunesFeed.books([feedType], [options]) ⇒ [<code>Promise.&lt;Feed&gt;</code>](#Feed)
Retrieves feed for Books.

**Kind**: instance method of [<code>ITunesFeed</code>](#ITunesFeed)  
**Returns**: [<code>Promise.&lt;Feed&gt;</code>](#Feed) - Instance of `Feed`.  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| [feedType] | <code>string</code> | <code>&quot;&#x27;top-free&#x27;&quot;</code> | The type of feed to retrieve. Must be one of: `top-free`, `top-paid`. |
| [options] | <code>Object</code> | <code>{}</code> | Any options from ITunesFeed to overwrite for this request. |

**Example**  
```js
let feed = await itunesFeed.books();
//=> [...]
```
<a name="ITunesFeed+iosApps"></a>

### itunesFeed.iosApps([feedType], [genre], [options]) ⇒ [<code>Promise.&lt;Feed&gt;</code>](#Feed)
Retrieves feed for iOS Apps.

**Kind**: instance method of [<code>ITunesFeed</code>](#ITunesFeed)  
**Returns**: [<code>Promise.&lt;Feed&gt;</code>](#Feed) - Instance of `Feed`.  
**Throws**:

- <code>Error</code> If an invalid genre is provided.
- <code>Error</code> If the genre is set to `games` on a feed type that is not `top-free` or `top-paid`.


| Param | Type | Default | Description |
| --- | --- | --- | --- |
| [feedType] | <code>string</code> | <code>&quot;&#x27;top-free&#x27;&quot;</code> | The type of feed to retrieve. Must be one of: `new-apps-we-love`, `new-games-we-love`, `top-free`, `top-free-ipad`, `top-grossing`, `top-grossing-ipad`, `top-paid`. |
| [genre] | <code>string</code> | <code>&quot;&#x27;all&#x27;&quot;</code> | Optional genre to retrieve. The only available genres are `all` and `games`. The `games` genre is only available on `top-free` and `top-paid` feed types. |
| [options] | <code>Object</code> | <code>{}</code> | Any options from ITunesFeed to overwrite for this request. |

**Example**  
```js
let feed = await itunesFeed.iosApps();
//=> [...]

let feed = await itunesFeed.iosApps('top-paid', 'games');
```
<a name="ITunesFeed+itunesMusic"></a>

### itunesFeed.itunesMusic([feedType], [options]) ⇒ [<code>Promise.&lt;Feed&gt;</code>](#Feed)
Retrieves feed for iTunes Music.

**Kind**: instance method of [<code>ITunesFeed</code>](#ITunesFeed)  
**Returns**: [<code>Promise.&lt;Feed&gt;</code>](#Feed) - Instance of `Feed`.  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| [feedType] | <code>string</code> | <code>&quot;&#x27;top-songs&#x27;&quot;</code> | The type of feed to retrieve. Must be one of: `new-music`, `recent-releases`, `top-albums`, `top-songs`. |
| [options] | <code>Object</code> | <code>{}</code> | Any options from ITunesFeed to overwrite for this request. |

**Example**  
```js
let feed = await itunesFeed.itunesMusic();
//=> [...]
```
<a name="ITunesFeed+itunesU"></a>

### itunesFeed.itunesU([options]) ⇒ [<code>Promise.&lt;Feed&gt;</code>](#Feed)
Retrieves feed for iTunes U.

**Kind**: instance method of [<code>ITunesFeed</code>](#ITunesFeed)  
**Returns**: [<code>Promise.&lt;Feed&gt;</code>](#Feed) - Instance of `Feed`.  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| [options] | <code>Object</code> | <code>{}</code> | Any options from ITunesFeed to overwrite for this request. |

**Example**  
```js
let feed = await itunesFeed.itunesU();
//=> [...]
```
<a name="ITunesFeed+macApps"></a>

### itunesFeed.macApps([feedType], [options]) ⇒ [<code>Promise.&lt;Feed&gt;</code>](#Feed)
Retrieves feed for MacOS Apps.

**Kind**: instance method of [<code>ITunesFeed</code>](#ITunesFeed)  
**Returns**: [<code>Promise.&lt;Feed&gt;</code>](#Feed) - Instance of `Feed`.  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| [feedType] | <code>string</code> | <code>&quot;&#x27;top-mac-apps&#x27;&quot;</code> | The type of feed to retrieve. Must be one of: `top-free-mac-apps`, `top-grossing-mac-apps`, `top-mac-apps`, `top-paid-mac-apps`. |
| [options] | <code>Object</code> | <code>{}</code> | Any options from ITunesFeed to overwrite for this request. |

**Example**  
```js
let feed = await itunesFeed.macApps();
//=> [...]
```
<a name="ITunesFeed+movies"></a>

### itunesFeed.movies([options]) ⇒ [<code>Promise.&lt;Feed&gt;</code>](#Feed)
Retrieves feed for Movies.

**Kind**: instance method of [<code>ITunesFeed</code>](#ITunesFeed)  
**Returns**: [<code>Promise.&lt;Feed&gt;</code>](#Feed) - Instance of `Feed`.  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| [options] | <code>Object</code> | <code>{}</code> | Any options from ITunesFeed to overwrite for this request. |

**Example**  
```js
let feed = await itunesFeed.movies();
//=> [...]
```
<a name="ITunesFeed+musicVideos"></a>

### itunesFeed.musicVideos([options]) ⇒ [<code>Promise.&lt;Feed&gt;</code>](#Feed)
Retrieves feed for Music Videos.

**Kind**: instance method of [<code>ITunesFeed</code>](#ITunesFeed)  
**Returns**: [<code>Promise.&lt;Feed&gt;</code>](#Feed) - Instance of `Feed`.  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| [options] | <code>Object</code> | <code>{}</code> | Any options from ITunesFeed to overwrite for this request. |

**Example**  
```js
let feed = await itunesFeed.musicVideos();
//=> [...]
```
<a name="ITunesFeed+podcasts"></a>

### itunesFeed.podcasts([options]) ⇒ [<code>Promise.&lt;Feed&gt;</code>](#Feed)
Retrieves feed for Podcasts.

**Kind**: instance method of [<code>ITunesFeed</code>](#ITunesFeed)  
**Returns**: [<code>Promise.&lt;Feed&gt;</code>](#Feed) - Instance of `Feed`.  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| [options] | <code>Object</code> | <code>{}</code> | Any options from ITunesFeed to overwrite for this request. |

**Example**  
```js
let feed = await itunesFeed.podcasts();
//=> [...]
```
<a name="ITunesFeed+tvShows"></a>

### itunesFeed.tvShows([feedType], [options]) ⇒ [<code>Promise.&lt;Feed&gt;</code>](#Feed)
Retrieves feed for TV Shows.

**Kind**: instance method of [<code>ITunesFeed</code>](#ITunesFeed)  
**Returns**: [<code>Promise.&lt;Feed&gt;</code>](#Feed) - Instance of `Feed`.  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| [feedType] | <code>string</code> | <code>&quot;&#x27;top-tv-episodes&#x27;&quot;</code> | The type of feed to retrieve. Must be one of: `top-tv-episodes`, `top-tv-seasons`. |
| [options] | <code>Object</code> | <code>{}</code> | Any options from ITunesFeed to overwrite for this request. |

**Example**  
```js
let feed = await itunesFeed.tvShows();
//=> [...]
```
<a name="Feed"></a>

## Feed ⇐ [<code>Array</code>](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array)
Data wrapper for feed responses. Returns an array of results from the feed with additional properties 
available on the prototype.

**Kind**: global class  
**Extends**: [<code>Array</code>](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array)  

* [Feed](#Feed) ⇐ [<code>Array</code>](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array)
    * [feed.title](#Feed+title) : <code>String</code>
    * [feed.url](#Feed+url) : <code>String</code>
    * [feed.countryCode](#Feed+countryCode) : <code>String</code>
    * [feed.updated](#Feed+updated) : <code>Date</code>

<a name="Feed+title"></a>

### feed.title : <code>String</code>
The title of the feed.

**Kind**: instance property of [<code>Feed</code>](#Feed)  
**Read only**: true  
<a name="Feed+url"></a>

### feed.url : <code>String</code>
The url of the feed.

**Kind**: instance property of [<code>Feed</code>](#Feed)  
**Read only**: true  
<a name="Feed+countryCode"></a>

### feed.countryCode : <code>String</code>
The country code in which this feed is for.

**Kind**: instance property of [<code>Feed</code>](#Feed)  
**Read only**: true  
<a name="Feed+updated"></a>

### feed.updated : <code>Date</code>
The date when this feed was last updated.

**Kind**: instance property of [<code>Feed</code>](#Feed)  
**Read only**: true  
## Tests
To run tests locally, clone this repo and run:

```
npm test
```

## License
MIT License


CHANGELOG.md
# 1.0.0 (2019-04-23)


### Bug Fixes

* check for e.response to be set when error handling ([c826f85](https://github.com/KyleRoss/node-itunes-feed/commit/c826f85))
* default feed.results to empty array if not available in the data ([e9e90fb](https://github.com/KyleRoss/node-itunes-feed/commit/e9e90fb))


### Features

* release 1.0.0 ([32254bc](https://github.com/KyleRoss/node-itunes-feed/commit/32254bc))


test/feed.js
"use strict";
const assert = require('assert');
const Feed = require('../lib/Feed');
const testFeed = require('./assets/test-feed.json');

describe('Feed', function() {
    let feed = new Feed(testFeed);
    
    it('should extend Array', function() {
        assert(feed instanceof Array);
    });
    
    it ('should have a length of 10', () => {
        assert.equal(feed.length, 10);
    });
    
    describe ('Getters', () => {
        it ('should have title', () => {
            assert.equal(typeof feed.title, 'string');
            assert.equal(feed.title, 'Top Courses');
        });
        
        it ('should have url', () => {
            assert.equal(typeof feed.url, 'string');
            assert.equal(feed.url, 'https://rss.itunes.apple.com/api/v1/us/itunes-u/top-itunes-u-courses/all/10/explicit.json');
        });
        
        it ('should have countryCode', () => {
            assert.equal(typeof feed.countryCode, 'string');
            assert.equal(feed.countryCode, 'us');
        });
        
        it ('should have updated', () => {
            assert.equal(typeof feed.updated, 'object');
            assert(feed.updated instanceof Date);
            assert.equal(feed.updated.toISOString(), '2019-04-03T09:59:08.000Z');
        });
    });
});


test/module.js
"use strict";
const assert = require('assert');
const itunesFeed = require('../');

describe('Module', function() {
    it('should return an instance of ITunesFeed', function() {
        assert.equal(itunesFeed.constructor.name, 'ITunesFeed');
    });
    
    it('should have access to uninstantiated class', function() {
        assert(typeof itunesFeed.ITunesFeed === 'function');
    });
});


test/itunesFeed.js
"use strict";
/* global AssertionError */
const assert = require('assert');
const ITunesFeed = require('../lib/ITunesFeed');
const Feed = require('../lib/Feed');
const { medias } = require('../lib/constants');

describe('ITunesFeed', function() {
    describe ('Class', () => {
        it('should have getter for uninstantiated class', () => {
            let inst = new ITunesFeed();
            assert.equal(typeof inst.ITunesFeed, 'function');
            assert.equal(inst.ITunesFeed.name, 'ITunesFeed');
        });
    });
    
    describe('Constructor', () => {
        it('should have default options', () => {
            let inst = new ITunesFeed();
            
            assert.equal(inst.options.countryCode, 'us');
            assert.equal(inst.options.limit, 100);
            assert.equal(inst.options.explicit, true);
        });
        
        it('should override default options', () => {
            let inst = new ITunesFeed({
                countryCode: 'al',
                limit: 50,
                explicit: false
            });
            
            assert.equal(inst.options.countryCode, 'al');
            assert.equal(inst.options.limit, 50);
            assert.equal(inst.options.explicit, false);
        });
        
        it ('should create an instance of axios', () => {
            let inst = new ITunesFeed();
            
            assert(inst._axios);
        });
    });
    
    describe ('Internal Methods', () => {
        describe ('_mergeOptions()', () => {
            let inst = new ITunesFeed();
            
            it('should merge options', () => {
                let res = inst._mergeOptions({ countryCode: 'al' });
                assert.equal(res.countryCode, 'al');
                assert.equal(res.limit, 100);
                assert.equal(res.explicit, true);
            });
            
            it ('should merge options into custom defaults', () => {
                let res = inst._mergeOptions({ countryCode: 'al' }, { limit: 50 });
                assert.equal(res.countryCode, 'al');
                assert.equal(res.limit, 50);
                assert.equal(res.explicit, undefined);
            });
            
            describe ('Errors', () => {
                it('should throw error when no country code is set', () => {
                    assert.throws(function() {
                        inst._mergeOptions({ countryCode: null }, { limit: 100 });
                    });
                });
                
                it('should throw error when country code is not 2 characters', () => {
                    assert.throws(function() {
                        inst._mergeOptions({ countryCode: 'abc' }, { limit: 100 });
                    });
                });
                
                it('should throw error when an invalid country code is set', () => {
                    assert.throws(function() {
                        inst._mergeOptions({ countryCode: 'zz' }, { limit: 100 });
                    });
                });
                
                it('should throw error when no limit is set', () => {
                    assert.throws(function() {
                        inst._mergeOptions({ limit: null }, { countryCode: 'us' });
                    });
                });
                
                it('should throw error when invalid limit is set', () => {
                    assert.throws(function() {
                        inst._mergeOptions({ limit: 200 }, { countryCode: 'us' });
                    });
                });
            });
        });
        
        describe ('_validateFeedType()', () => {
            let inst = new ITunesFeed();
            
            it('should return true when feed type is matched', () => {
                assert.equal(inst._validateFeedType(['test'], 'test'), true);
            });
            
            it ('should throw error if type is not provided', () => {
                assert.throws(function() {
                    inst._validateFeedType([]);
                });
            });
            
            it ('should throw error when no feed type is matched', () => {
                assert.throws(function() {
                    inst._validateFeedType(['abc'], ['123']);
                });
            });
        });
        
        describe ('_generateFeedURL()', function() {
            let inst = new ITunesFeed();
            
            it('should generate an explicit feed url', async () => {
                let res = await inst._generateFeedURL('us', 'apple-music', 'top-songs', 'all', 50, true);
                assert.equal(res, 'https://rss.itunes.apple.com/api/v1/us/apple-music/top-songs/all/50/explicit.json');
            });
            
            it('should generate an non-explicit feed url', async () => {
                let res = await inst._generateFeedURL('us', 'apple-music', 'top-songs', 'all', 50, false);
                assert.equal(res, 'https://rss.itunes.apple.com/api/v1/us/apple-music/top-songs/all/50/non-explicit.json');
            });
        });
        
        describe ('_getFeed()', function() {
            this.timeout(5000);
            let inst = new ITunesFeed({
                limit: 10
            });
            
            it('should return an instance of Feed', async () => {
                let res = await inst._getFeed(medias.APPLE_MUSIC, 'top-songs', 'all');
                assert(res instanceof Feed);
            });
            
            it('should throw an error with an improper feed', async () => {
                let testMedia = {
                    mediaType: 'test',
                    types: ['test']
                };
                
                try {
                    await inst._getFeed(testMedia, 'test', 'all');
                } catch(e) {
                    if(e && e instanceof Error) {
                        assert.equal(e.message, 'iTunes Feed not found. Please check your parameters.');
                        return true;
                    }
                }
                
                throw new AssertionError('Did not throw an error with an improper feed');
            });
        });
    });
    
    describe('Methods', function() {
        this.timeout(5000);
        let inst = new ITunesFeed({
            limit: 25
        });
        
        describe ('appleMusic()', () => {
            it('should return an instance of Feed with default options', async () => {
                let res = await inst.appleMusic();
                assert(res instanceof Feed);
            });
            
            it ('should return an instance of feed with custom options', async () => {
                let res = await inst.appleMusic('hot-tracks', { limit: 10 });
                assert(res instanceof Feed);
            });
        });
        
        describe ('audioBooks()', () => {
            it('should return an instance of Feed with default options', async () => {
                let res = await inst.audioBooks();
                assert(res instanceof Feed);
            });
            
            it ('should return an instance of feed with custom options', async () => {
                let res = await inst.audioBooks({ limit: 10 });
                assert(res instanceof Feed);
            });
        });
        
        describe ('books()', () => {
            it('should return an instance of Feed with default options', async () => {
                let res = await inst.books();
                assert(res instanceof Feed);
            });
            
            it ('should return an instance of feed with custom options', async () => {
                let res = await inst.books('top-paid', { limit: 10 });
                assert(res instanceof Feed);
            });
        });
        
        describe ('iosApps()', () => {
            it('should return an instance of Feed with default options', async () => {
                let res = await inst.iosApps();
                assert(res instanceof Feed);
            });
            
            it ('should return an instance of feed with custom options', async () => {
                let res = await inst.iosApps('top-paid', 'games', { limit: 10 });
                assert(res instanceof Feed);
            });
            
            it ('should throw an error if invalid genre is provided', async () => {
                try {
                    await inst.iosApps('top-paid', 'not-valid', { limit: 10 });
                } catch(e) {
                    if(e && e instanceof Error) return true;
                }
                
                throw new AssertionError('Did not throw an error for an invalid genre');
            });
            
            it ('should throw an error if invalid genre is provided for given feed type', async () => {
                try {
                    await inst.iosApps('top-paid-ipad', 'games', { limit: 10 });
                } catch(e) {
                    if(e && e instanceof Error) return true;
                }
                
                throw new AssertionError('Did not throw an error for an invalid genre');
            });
        });
        
        describe ('itunesMusic()', () => {
            it('should return an instance of Feed with default options', async () => {
                let res = await inst.itunesMusic();
                assert(res instanceof Feed);
            });
            
            it ('should return an instance of feed with custom options', async () => {
                let res = await inst.itunesMusic('new-music', { limit: 10 });
                assert(res instanceof Feed);
            });
        });
        
        describe ('itunesU()', () => {
            it('should return an instance of Feed with default options', async () => {
                let res = await inst.itunesU();
                assert(res instanceof Feed);
            });
            
            it ('should return an instance of feed with custom options', async () => {
                let res = await inst.itunesU({ limit: 10 });
                assert(res instanceof Feed);
            });
        });
        
        describe ('macApps()', () => {
            it('should return an instance of Feed with default options', async () => {
                let res = await inst.macApps();
                assert(res instanceof Feed);
            });
            
            it ('should return an instance of feed with custom options', async () => {
                let res = await inst.macApps('top-free-mac-apps', { limit: 10 });
                assert(res instanceof Feed);
            });
        });
        
        describe ('movies()', () => {
            it('should return an instance of Feed with default options', async () => {
                let res = await inst.movies();
                assert(res instanceof Feed);
            });
            
            it ('should return an instance of feed with custom options', async () => {
                let res = await inst.movies({ limit: 10 });
                assert(res instanceof Feed);
            });
        });
        
        describe ('musicVideos()', () => {
            it('should return an instance of Feed with default options', async () => {
                let res = await inst.musicVideos();
                assert(res instanceof Feed);
            });
            
            it ('should return an instance of feed with custom options', async () => {
                let res = await inst.musicVideos({ limit: 10 });
                assert(res instanceof Feed);
            });
        });
        
        describe ('podcasts()', () => {
            it('should return an instance of Feed with default options', async () => {
                let res = await inst.podcasts();
                assert(res instanceof Feed);
            });
            
            it ('should return an instance of feed with custom options', async () => {
                let res = await inst.podcasts({ limit: 10 });
                assert(res instanceof Feed);
            });
        });
        
        describe ('tvShows()', () => {
            it('should return an instance of Feed with default options', async () => {
                let res = await inst.tvShows();
                assert(res instanceof Feed);
            });
            
            it ('should return an instance of feed with custom options', async () => {
                let res = await inst.tvShows('top-tv-seasons', { limit: 10 });
                assert(res instanceof Feed);
            });
        });
    });
});


lib/Feed.js
"use strict";
/**
 * @external Array
 * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array
 */

/**
 * Data wrapper for feed responses. Returns an array of results from the feed with additional properties 
 * available on the prototype.
 * @extends {external:Array}
 * @typicalname feed
 */
class Feed extends Array {
    constructor({ feed }) {
        super(...feed.results);
        this._feed = feed;
    }
    
    /**
     * The title of the feed.
     * @readonly
     * @type {String}
     */
    get title() {
        return this._feed.title;
    }
    
    /**
     * The url of the feed.
     * @readonly
     * @type {String}
     */
    get url() {
        return this._feed.id;
    }
    
    /**
     * The country code in which this feed is for.
     * @readonly
     * @type {String}
     */
    get countryCode() {
        return this._feed.country;
    }
    
    /**
     * The date when this feed was last updated.
     * @readonly
     * @type {Date}
     */
    get updated() {
        return new Date(this._feed.updated);
    }
}

module.exports = Feed;


lib/constants.js
"use strict";

module.exports = {
    medias: {
        APPLE_MUSIC: {
            mediaType: 'apple-music',
            types: [
                'coming-soon',
                'hot-tracks',
                'new-releases',
                'top-albums',
                'top-songs'
            ]
        },
        AUDIOBOOKS: {
            mediaType: 'audiobooks',
            types: ['top-audiobooks']
        },
        BOOKS: {
            mediaType: 'books',
            types: [
                'top-free',
                'top-paid'
            ]
        },
        IOS_APPS: {
            mediaType: 'ios-apps',
            types: [
                'new-apps-we-love',
                'new-games-we-love',
                'top-free',
                'top-free-ipad',
                'top-grossing',
                'top-grossing-ipad',
                'top-paid'
            ]
        },
        ITUNES_MUSIC: {
            mediaType: 'itunes-music',
            types: [
                'new-music',
                'recent-releases',
                'top-albums',
                'top-songs'
            ]
        },
        ITUNES_U: {
            mediaType: 'itunes-u',
            types: ['top-itunes-u-courses']
        },
        MAC_APPS: {
            mediaType: 'macos-apps',
            types: [
                'top-free-mac-apps',
                'top-grossing-mac-apps',
                'top-mac-apps',
                'top-paid-mac-apps'
            ]
        },
        MOVIES: {
            mediaType: 'movies',
            types: ['top-movies']
        },
        MUSIC_VIDEOS: {
            mediaType: 'music-videos',
            types: ['top-music-videos']
        },
        PODCASTS: {
            mediaType: 'podcasts',
            types: ['top-podcasts']
        },
        TV_SHOWS: {
            mediaType: 'tv-shows',
            types: [
                'top-tv-episodes',
                'top-tv-seasons'
            ]
        }
    },
    
    
    countries: [
        "al",
        "dz",
        "ao",
        "ai",
        "ag",
        "ar",
        "am",
        "au",
        "at",
        "az",
        "bs",
        "bh",
        "bb",
        "by",
        "be",
        "bz",
        "bj",
        "bm",
        "bt",
        "bo",
        "bw",
        "br",
        "vg",
        "bn",
        "bg",
        "bf",
        "kh",
        "ca",
        "cv",
        "ky",
        "td",
        "cl",
        "cn",
        "co",
        "cg",
        "cr",
        "hr",
        "cy",
        "cz",
        "dk",
        "dm",
        "do",
        "ec",
        "eg",
        "sv",
        "ee",
        "fj",
        "fi",
        "fr",
        "gm",
        "de",
        "gh",
        "gr",
        "gd",
        "gt",
        "gw",
        "gy",
        "hn",
        "hk",
        "hu",
        "is",
        "in",
        "id",
        "ie",
        "il",
        "it",
        "jm",
        "jp",
        "jo",
        "kz",
        "ke",
        "kr",
        "kw",
        "kg",
        "la",
        "lv",
        "lb",
        "lr",
        "lt",
        "lu",
        "mo",
        "mk",
        "mg",
        "mw",
        "my",
        "ml",
        "mt",
        "mr",
        "mu",
        "mx",
        "fm",
        "md",
        "mn",
        "ms",
        "mz",
        "na",
        "np",
        "nl",
        "nz",
        "ni",
        "ne",
        "ng",
        "no",
        "om",
        "pk",
        "pw",
        "pa",
        "pg",
        "py",
        "pe",
        "ph",
        "pl",
        "pt",
        "qa",
        "ro",
        "ru",
        "sa",
        "sn",
        "sc",
        "sl",
        "sg",
        "sk",
        "si",
        "sb",
        "za",
        "es",
        "lk",
        "kn",
        "lc",
        "vc",
        "sr",
        "sz",
        "se",
        "ch",
        "st",
        "tw",
        "tj",
        "tz",
        "th",
        "tt",
        "tn",
        "tr",
        "tm",
        "tc",
        "ug",
        "ua",
        "ae",
        "gb",
        "us",
        "uy",
        "uz",
        "ve",
        "vn",
        "ye",
        "zw"
    ],
    
    
    limits: [
        10,
        25,
        50,
        100
    ]
};
